core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java

```diff
@@ -112,51 +97,41 @@ public final void reroute(String reason) {
     @Override
     public void clusterChanged(ClusterChangedEvent event) {
         if (event.state().nodes().localNodeMaster()) {
-            // figure out when the next unassigned allocation need to happen from now. If this is larger or equal
-            // then the last time we checked and scheduled, we are guaranteed to have a reroute until then, so no need
+            // figure out when the next unassigned allocation needs to happen from now. If this is larger or equal
+            // than the last time we checked and scheduled, we are guaranteed to have a reroute until then, so no need
             // to schedule again
-            long nextDelaySetting = UnassignedInfo.findSmallestDelayedAllocationSetting(settings, event.state());
-            if (nextDelaySetting > 0 && nextDelaySetting < registeredNextDelaySetting) {
+            long nextDelaySettingMillis = UnassignedInfo.findSmallestDelayedAllocationSetting(settings, event.state());
```

Comment:
I got confused reading this again and spent a good 10 minutes convincing myself, again, that this is a safe way to decide re-scheduling or not. Here is what I think will make it clearer:

1) The comment should say: "figure out if an existing scheduled reroute is good enough or whether we need to cancel and reschedule. If the the current minimum  of the currently relevant delay settings is larger than something we scheduled in the past, we are guaranteed a the planned schedule will happen before any of the current shard delays are expired."
2) rename nextDelaySetting to minDelaySettings.
3) rename registeredNextDelaySetting to minDelaySettingsAtLastScheduling.
4) Only set minDelaySettingsAtLastScheduling if we are actually going to schedule and restructure the if:

```
if (minDelaySettings <=0) {
  logger.trace("no need to schedule reroute - no delayed unassigned shards, minDelaySettings [{}], scheduled [{}]", minDelaySettings, minDelaySettingsAtLastScheduling);
   minDelaySettingsAtLastScheduling = Long.MAX_VALUE;
   FutureUtils.cancel(registeredNextDelayFuture);.
} else if (minDelaySettings <  minDelaySettingsAtLastScheduling) {
    FutureUtils.cancel(registeredNextDelayFuture);
    minDelaySettingsAtLastScheduling = minDelaySettings;
   TimeValue nextDelay = TimeValue.timeValueNanos(UnassignedInfo.findNextDelayedAllocationIn(event.state()));
   assert nextDelay > 0 : "next delay must be non 0 as minDelaySettings is [" + minDelaySettings + "];
   logger.info("delaying allocation for [{}] unassigned shards, next check in [{}]",
                             unassignedDelayedShards, nextDelay);
   registeredNextDelayFuture = threadPool.schedule(nextDelay ...
} else {
    logger.trace("no need to schedule reroute - current schedule reroute is enough. minDelaySettings [{}], scheduled [{}]", 
}
```